tracer/src/Datadog.Trace/DuckTyping/DuckType.cs

```diff
@@ -155,20 +175,35 @@ private static CreateTypeResult CreateProxyType(Type proxyDefinitionType, Type t
             //     }
             // }
 
-            lock (_locker)
+            lock (Locker)
             {
                 try
                 {
-                    var moduleBuilder = CreateTypeAndModuleBuilder(proxyDefinitionType, targetType, out var proxyTypeBuilder, out var instanceField);
+                    ModuleBuilder? moduleBuilder = null;
+                    TypeBuilder? proxyTypeBuilder = null;
+                    FieldInfo? instanceField = null;
+
+                    if (!dryRun)
+                    {
+                        moduleBuilder = CreateTypeAndModuleBuilder(proxyDefinitionType, targetType, out proxyTypeBuilder, out instanceField);
+                    }
 
                     if (proxyDefinitionType.IsValueType)
                     {
                         // Create Fields and Properties from the struct information
                         CreatePropertiesFromStruct(proxyTypeBuilder, proxyDefinitionType, targetType, instanceField);
 
-                        // Create Type
-                        Type proxyType = proxyTypeBuilder.CreateTypeInfo().AsType();
-                        return new CreateTypeResult(proxyDefinitionType, proxyType, targetType, CreateStructCopyMethod(moduleBuilder, proxyDefinitionType, proxyType, targetType), null);
+                        if (proxyTypeBuilder is not null)
+                        {
+                            // Create Type
+                            Type? proxyType = proxyTypeBuilder.CreateTypeInfo()?.AsType();
+                            if (proxyType is not null)
```

Comment:
if proxyType _is_ `null` here, we create a `CreateTypeResult` without an activator _or_ an exceptionInfo. That seems confusing to me `CreateTypeResult` should be a discriminated union of Success+Activator or Failure+Exception 🤔 Seems like we should be adding exception info here in this case (or if `proxyTypeBuilder` is null), even if it's default/shared?

Reply:
I've changed this here: [1d5d28b](https://github.com/DataDog/dd-trace-dotnet/pull/2716/commits/1d5d28bb11abc009955dc327b1ff02da344f6d4a) and now we throw if the proxyTypeBuilder.CreateTypeInfo() is null, to ensure that exceptionInfo.